EIPS/eip-5450.md

```diff
@@ -13,51 +13,100 @@ requires: 3540, 3670, 4200, 4750
-2. For each instruction, data stack always has enough items, i.e. stack underflow is invalid.
-3. For `CALLF` instruction, data stack has enough items to use as input arguments to a called function according to its type defined in the type section.
-4. For `RETF` instruction, data stack before executing it has exactly `n` items to use as output values returned from a function, where `n` is function's number of outputs according to its type defined in the type section.
-5. `max_stack_height` does not exceed 1023.
-6. The maximum data stack height matches the corresponding code section's `max_stack_height` within the type section body.
-7. No unreachable instructions exist in the code section.
+For each reachable instruction in the code the operand stack height is recorded.
+The first instruction has the recorded stack height equal to the number of inputs to the function type matching the code (`type[code_section_index].inputs`).
 
-To examine every reachable code path, validation needs to traverse every instruction in order, while also following each non-conditional jump, and following both possible branches for each conditional jump. After completing this, verify each instruction was visited at least once. Fail if any instructions were not visited as this invalidates 7).
+The FIFO queue *worklist* is used for tracking "to be inspected" instructions. Initially, it contains the first instruction.
 
-The complexity of this traversal is linear in the number of instructions, because each code path is examined only once, and property 1 guarantees no loops in the validation.
+While *worklist* is not empty, dequeue an instruction and:
+
+1. Determine the effect the instruction has on the operand stack:
+   1. **Check** if the recorded stack height satisfies the instruction requirements. Specifically:
+      - for `CALLF` instruction the required stack height is the number of inputs of the called function according to its type defined in the type section.
+      - for `RETF` instruction the recorded stack height must be exactly the number of outputs of the function matching the code.
+   2. Compute new stack height after the instruction execution.
+2. Determine the list of successor instructions than can follow the current instructions:
+   1. The next instruction for all instructions other than terminating instructions and unconditional jump.
+   2. All targets of a conditional or unconditional jump.
+3. For each successor instruction:
+   1. **Check** if the instruction is present in the code (i.e. execution must not "fall off" the code).
+   2. If the instruction does not have stack height recorded (visited for the first time):
+      1. Record the instruction stack height as the value computed in 1.2.
+      2. Add the instruction to the *worklist*.
+   3. Otherwise, **check** if the recorded stack height equals the value computed in 1.2.
+4. **Check** if all instruction where reached by the analysis.
```

Comment:
This looks like it's part if `While *worklist* is not empty` loop. Maybe make it a separate section, like 
```

When *worklist* is empty:

4. **Check** if all instruction where reached by the analysis.
5. ...
```

Reply:
Done.